### ISSUE TITLE
Can you have a little explain about the a0 parameter and rate in sgd part

### ISSUE THREAD

**Issue by fooSynaptic** [timvieira/crf#14]
Really apperciate your work, i have investigated your code but have some confusion part.
in the code
```python
    def sgd(self, data, iterations=20, a0=10, validate=None):
        """ Parameter estimation with stochastic gradient descent (sgd). """
        self.preprocess(data)
        W = self.W
        for i in range(iterations):
            rate = a0 / (i + 1)**0.501
            for x in iterview(data, msg='Iteration %s' % i):
                for k, v in self.expectation(x).items():
                    W[k] -= rate*v
                for k in x.target_features:
                    W[k] += rate
            if validate:
                validate(self, i)
```
- Is it a parameter restore or fix for the warm-up stage? for the result the gradient at warm-up stage may fluctuate violently, so restore parameter by add some scales can mitigate this.
- But why not fix the gradient(here is the expectation logit) directly, is there some theory analysis about these two method?

**Comment by fooSynaptic** [comment]
or it's just a learning rate decay convention implementation?

**Comment by timvieira** [comment]
```
rate = a0 / (i + 1)**0.501
```
This is a standard learning-rate decay for stochastic gradient.

Stochastic gradient methods allow us to scale to large datasets.

**Comment by fooSynaptic** [comment]
Sorry, i just cannot understand why the `W[k]` need add `rate` in every update step(to my understand, the k here is the label dimension). Any formulation can explain this implement?

**Comment by timvieira** [comment]
`k` is the feature index, not the label.

**Comment by fooSynaptic** [comment]
thanks
